### PR TITLE
(Re-merge after sync with slime) [Feat] Add LoRA for FSDP backend. (#307)

### PR DESCRIPTION
**Pull request by GuanxingLu**
WIP with @PopSoda2002. 

Approach: Save adapter to disk -> rollout engine loads via load_lora_adapter.

Limitation: When rollout engine offloading is enabled, the base model weight should be re-transmitted in each update (though overhead is manageable).

TODOs:
- [x] Benchmark Reward curve.
- [x] Benchmark memory usage.
- [ ] Test LoRA on MoE models.
- [ ] Optimize LoRA weight update speed.
- [ ] Optimize LoRA rollout speed.

Here is the training reward curve of Qwen-4B with (pink) and without (green) LoRA:
image
For complete results, please refer to this W&B report: https://api.wandb.ai/links/747398423/i3u8ebrj

### CODE REVIEW

**Comment by PopSoda2002** [comment]
I also test this PR, for 1 H100 and model Qwen3-4B
If I use LoRA(rank=1) it will not OOM:
image
If I do not use LoRA, it will OOM:
image

File: miles/backends/fsdp_utils/lora_utils.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+
+def is_lora_model(module: nn.Module) -> bool:
+    unwrapped = getattr(module, "_fsdp_wrapped_module", module)
+    return hasattr(unwrapped, "peft_config")
+
+
+def save_lora_to_disk(module: nn.Module, save_dir: str) -> str:
+    """Save LoRA adapter to disk with file lock mechanism."""
+    options = StateDictOptions(full_state_dict=True, cpu_offload=True)
```
Comment (PopSoda2002) [inline miles/backends/fsdp_utils/lora_utils.py:54]: Can you add a TODO here? TODO: All gather lora layers not full layers

File: miles/backends/fsdp_utils/lora_utils.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    full_state_dict = get_model_state_dict(module, options=options)
+
+    state_dict = {name: param for name, param in full_state_dict.items() if "lora_" in name}
```
Comment (PopSoda2002) [inline miles/backends/fsdp_utils/lora_utils.py]: nit: change the name from `state_dict` to `lora_state_dict`

File: miles/utils/arguments.py [hunk 3]
```diff
@@ -104,6 +104,15 @@ def add_cluster_arguments(parser):
                     "This will always be true when --colocate is set."
                 ),
             )
+            parser.add_argument(
+                "--offload-rollout-level",
+                type=int,
+                default=2,
```
Comment (PopSoda2002) [inline miles/utils/arguments.py:110]: I am not sure if we need a more descriptive name like: `offload-rollout-kvcache-only`? But it's fine

File: train.py [hunk 4]
```diff
@@ -68,7 +68,16 @@ def onload_rollout():
         rollout_data_ref = ray.get(rollout_manager.generate.remote(rollout_id))
 
         if args.offload_rollout:
-            ray.get(rollout_manager.offload.remote())
+            # level 1: offload kv cache only, level 2: offload weights + kv cache
```
Comment (PopSoda2002) [inline train.py:71]: Can you add a TODO like: TODO: Optimization in the future: offload model weights to cpu to make more space for training?

**Comment by zhaochenyang20** [comment]
Blocked by a PR of Lora in SGLang

**Comment by PopSoda2002** [comment]
@GuanxingLu Hi guanxing, plz fix the pre-commit if you are free!

**Comment by GuanxingLu** [comment]
> @GuanxingLu Hi guanxing, plz fix the pre-commit if you are free!

Fixed, sorry for the oversight.

**Comment by emergenz** [comment]
Hi, I know there is likely no other solution before SGLang addresses [this](https://github.com/sgl-project/sglang/issues/15703), but did you quickly profile whether disk-based adapter reload is a bottleneck?
